### PR TITLE
JS: Add model for puppeteer

### PR DESCRIPTION
**Pull request by erik-krogh**
Recognizes the sink in CVE-2020-8902. 

Adds model for `puppeteer`. 

And adds `DataFlow::SourceNode::hasUnderlyingType(_, _)` as API-nodes.   
These new nodes are accessible through the `API::Node::ofType` predicate (and are not reachable from `API::root()`). 
/ping @max-schaefer: what do you think? 

Evaluations look good: [one](https://github.com/dsp-testing/erik-krogh-dca/issues/34), [two](https://github.com/dsp-testing/erik-krogh-dca/tree/run/fogery-default-security/reports).  
No change in performance or results.

### CODE REVIEW

**Comment by max-schaefer** [comment]
> These new nodes are accessible through the `API::Node::ofType` predicate (and are not reachable from `API::root()`).
> /ping @max-schaefer: what do you think?

Can you briefly comment on why we can't represent this the way we previously did it for other TypeScript types, i.e., have an edge from the root node to a ref node for the type, and then an instance edge from there to the nodes that have this type?

Also, this PR seems like a good opportunity to try out the new `JS:changes-sources-or-sinks` label.

**Comment by erik-krogh** [comment]
> Can you briefly comment on why we can't represent this the way we previously did it for other TypeScript types, i.e., have an edge from the root node to a ref node for the type, and then an instance edge from there to the nodes that have this type?

We can (I just did the refactorization). But I don't think it makes a difference.

**Comment by max-schaefer** [comment]
> I'm still missing the edge from root.

Right, I would have expected there to be an edge from root to the canonical name of the type. It probably doesn't matter for the end user, but would be nice to figure out why it's not there.

**Comment by erik-krogh** [comment]
I got an edge from root now.  
I'm using `MkModuleImport` to represent the "module" of the TypeAnnotation, even though this is not entirely correct.  
An alternative is to directly add a `moduleName.exprotsName` edge from the root.

File: javascript/ql/src/semmle/javascript/ApiGraphs.qll [hunk 1]
```diff
@@ -413,6 +417,13 @@ module API {
         not n.isRoot() and
         isUsed(n)
       } or
+      /**
+       * A TypeScript type, identified by name of the type-annotation.
+       * This API node is exclusively used by `API::Node::ofType`.
+       */
+      MkHasUnderlyingType(string moduleName, string exportName) {
+        any(TypeAnnotation n).hasQualifiedName(moduleName, exportName)
```
Comment (asgerf) [inline javascript/ql/src/semmle/javascript/ApiGraphs.qll:425]: We could use `any(Type t).hasUnderlyingType` in addition to the above, for cases where the value comes in through a callback parameter without a type annotation, but where the type is inferred by contextual typing.

Or will it generate too many unneeded API nodes?

**Comment by max-schaefer** [comment]
Could we perhaps reuse `MkCanonicalNameUse` for this? It currently requires there to be a canonical name for the type, but that's more for restricting the argument than anything else, so if there isn't one available here we can relax the restrictions on that.

**Comment by erik-krogh** [comment]
> Could we perhaps reuse `MkCanonicalNameUse` for this? It currently requires there to be a canonical name for the type, but that's more for restricting the argument than anything else, so if there isn't one available here we can relax the restrictions on that.

I don't think we can. I don't see how we can combine them both into `MkCanonicalNameUse`.

**Review by asgerf — COMMENTED** [review]
+💯  - I've wanted to make this change and it's been nagging me that I couldn't get around to it

Does change make the `CanonicalName` nodes redundant? I believe they were added to support this exact use-case but the `hasUnderlyingType` node now has that covered.

**Comment by max-schaefer** [comment]
> Does this change make the CanonicalName nodes redundant?

That would be great, too! I'd prefer not to have two solutions for very similar problems if we can avoid it.

**Comment by erik-krogh** [comment]
> Does this change make the CanonicalName nodes redundant? I believe they were added to support this exact use-case but the hasUnderlyingType node now has that covered.

With your suggested `any(Type t).hasUnderlyingType` addition, then yes, I think the `CanonicalName` nodes are redundant.  
I tried to just delete them, and all the tests passed, so I'll commit it.

**Comment by asgerf** [comment]
Thanks! Do you want to run another evaluation?

**Comment by erik-krogh** [comment]
> Thanks! Do you want to run another evaluation?

[Here it is](https://github.com/dsp-testing/erik-krogh-dca/tree/run/fogery-defAndTS-security/reports).   

For some reason `TypeScript` kept failing (also on `main`).  
But it works fine on my machine :shrug:

**Review by asgerf — APPROVED** [review]

**Comment by erik-krogh** [comment]
@max-schaefer: any comments?

**Comment by max-schaefer** [comment]
Sorry, I'm a bit swamped today. Feel free to merge, I'll take a look later.

File: javascript/ql/src/semmle/javascript/ApiGraphs.qll [hunk 2]
```diff
@@ -384,6 +381,8 @@ module API {
         imports(_, m)
         or
         m = any(CanonicalName n | isUsed(n)).getExternalModuleName()
+        or
+        any(TypeAnnotation n).hasQualifiedName(m, _)
```
Comment (max-schaefer) [inline javascript/ql/src/semmle/javascript/ApiGraphs.qll:385]: Do we want to add `or any(Type t).hasUnderlyingType(m, _)` here as well?

**Review by max-schaefer — COMMENTED** [review]
Just one quick comment so far.

I am a little uneasy about having lost the distinction between type definitions and uses, but that doesn't matter when using API graphs for library modelling. (And apparently the tests I wrote aren't good enough to flag up any problems it might cause with flow summaries.)

File: javascript/ql/src/semmle/javascript/ApiGraphs.qll (hunk 2 above)
Comment (erik-krogh) [reply javascript/ql/src/semmle/javascript/ApiGraphs.qll:385]: Yes we do. And the remaining references to `CannonicalName` can be deleted.  
https://github.com/github/codeql/pull/5460

**Pull request merged**